### PR TITLE
Ensure correct attack damage on slot change

### PR DESCRIPTION
**Pull request by ThatApplePieGuy**
fixes a vanilla bug where attack damage isnt updated in time if you attack too fast after changing slots. demo: https://youtu.be/QyarG6DSh1c

### CODE REVIEW

**Review by hpfxd — APPROVED** [review]

**Review by uRyanxD — APPROVED** [review]
LGTM, however, I don't have time to test the patch

File: patches/server/0098-Ensure-correct-attack-damage-on-slot-change.patch [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
++        this.inventory.prevItemInHandIndex = this.inventory.itemInHandIndex; // PandaSpigot
+         this.inventory.itemInHandIndex = nbttagcompound.getInt("SelectedItemSlot");
+         this.sleeping = nbttagcompound.getBoolean("Sleeping");
+         this.sleepTicks = nbttagcompound.getShort("SleepTimer");
+@@ -957,6 +958,16 @@ public abstract class EntityHuman extends EntityLiving {
+     public void attack(Entity entity) {
+         if (entity.aD()) {
+             if (!entity.l(this)) {
++                // PandaSpigot start - ensure attack damage attributes from previous held item are removed
++                if (this.inventory.items[this.inventory.prevItemInHandIndex] != null) {
```
Comment (MachineBreaker) [inline patches/server/0098-Ensure-correct-attack-damage-on-slot-change.patch:24]: Make this a single array access

**Review by MachineBreaker — CHANGES_REQUESTED** [review]

**Comment by andreasdc** [comment]
Better change this in held item packet

**Review by MachineBreaker — APPROVED** [review]

**Comment by bl4stySKID** [comment]
this breaks what is known as "rod tricking" which is a pretty used mechanic on 1.8 for anyone wondering

**Comment by iamnoksio** [comment]
[00:41:06] [Server thread/WARN]: Failed to handle packet for /199.102.202.174:59257
java.lang.IllegalArgumentException: Modifier is already applied on this attribute!
	at net.minecraft.server.v1_7_R4.AttributeModifiable.a(SourceFile:74) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.PlayerConnection.a(PlayerConnection.java:847) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.PacketPlayInHeldItemSlot.a(SourceFile:30) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.PacketPlayInHeldItemSlot.handle(SourceFile:8) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.NetworkManager.a(NetworkManager.java:197) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.ServerConnection.c(ServerConnection.java:73) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.MinecraftServer.v(MinecraftServer.java:799) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.DedicatedServer.v(DedicatedServer.java:319) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.MinecraftServer.u(MinecraftServer.java:652) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.MinecraftServer.run(MinecraftServer.java:566) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.ThreadServerApplication.run(SourceFile:628) ~[spigot.jar:git-PaperSpigot-1f7d532]

**Comment by iamnoksio** [comment]
[00:41:06] [Server thread/WARN]: Failed to handle packet for /
java.lang.IllegalArgumentException: Modifier is already applied on this attribute!
	at net.minecraft.server.v1_7_R4.AttributeModifiable.a(SourceFile:74) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.PlayerConnection.a(PlayerConnection.java:847) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.PacketPlayInHeldItemSlot.a(SourceFile:30) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.PacketPlayInHeldItemSlot.handle(SourceFile:8) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.NetworkManager.a(NetworkManager.java:197) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.ServerConnection.c(ServerConnection.java:73) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.MinecraftServer.v(MinecraftServer.java:799) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.DedicatedServer.v(DedicatedServer.java:319) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.MinecraftServer.u(MinecraftServer.java:652) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.MinecraftServer.run(MinecraftServer.java:566) ~[spigot.jar:git-PaperSpigot-1f7d532]
	at net.minecraft.server.v1_7_R4.ThreadServerApplication.run(SourceFile:628) ~[spigot.jar:git-PaperSpigot-1f7d532]

**Comment by ThatApplePieGuy** [comment]
can you reproduce this in 1.8?

**Pull request closed without merging**